docs/source/tutorial_worlds.md

```diff
@@ -13,120 +13,353 @@ section. If you are interested in creating a task, please see
+0  |   1    |  Frankly, my dear, I don't give a damn. | 8
+1  |   2    |  I'm going to make him an offer he can't refuse. |      10
+7  |   2    |  You are without a doubt the worst pirate I've ever heard of.  |      12
+9  |   2    |  A million dollars isn't cool. You know what's cool? A billion dollars. |      12
+10 |   3    |  I'm as mad as hell, and I'm not going to take this anymore! |      13
+2  |   3    |  Mama always said life was like a box of chocolates. You never know what you're gonna get. |      17
+5  |   3    |  Do you wish me a good morning, or mean that it is a good morning whether I want it or not; or that you feel good this morning; or that it is a morning to be good on? |      38
+
+This is called the __batchsort__ algorithm. Note that our batch size remains
+fixed at 4, but we've grouped our worst examples so that they happen at the
+same time.
+
+:::{tip}
+You can use this your training or evaluation runs with `-dynb batchsort` or
+`--dynamic-batching batchsort`.
+:::
+
+::{note}
+Notice how the conversations no longer play in order. ParlAI handles all of this
+extra complexity for you. Remember, you only need to implement state tracking
+for _individual_ conversations, and along with `batch\_act`.
+::
+
+But we can take this even one step further, and implement fully dynamic batching.
+In dynamic batching, we grow our batches so that the number of _words_ per batch
+stays relatively fixed. This means we can process many short conversations at
+the same time. Let's imagine we set the number of total number of words per batch
+to be 80.
+
+ID | Batch #|  Message | # Words
+---|--------|----------|-------
+5  |   0    |  Do you wish me a good morning, or mean that it is a good morning whether I want it or not; or that you feel good this morning; or that it is a morning to be good on? |      38
+2  |   0    |  Mama always said life was like a box of chocolates. You never know what you're gonna get. |      17
+10 |   0    |  I'm as mad as hell, and I'm not going to take this anymore! |      13
+9  |   0    |  A million dollars isn't cool. You know what's cool? A billion dollars. |      12
+7  |   1    |  You are without a doubt the worst pirate I've ever heard of.  |      12
+1  |   1    |  I'm going to make him an offer he can't refuse. |      10
+0  |   1    |  Frankly, my dear, I don't give a damn. | 8
+6  |   1    |  I'm walking here! I'm walking here! |       6
+8  |   1    |  May the Force be with you. |       6
+11 |   1    |  I'm king of the world! |       5
+3  |   1    |  Here's Johnny! |       2
+4  |   1    |  Expecto patronum! | 2
+
+With this algorithm, we tried to get as close to 80 words as possible, without
+going over. The first batch ends up with exactly 80 words. The batch ends up
+with only 51. But we've now reduced the number of batches we need to process
+from 4 to only 2! This is the trick of how dynamic batching can provide
+[massive speed ups](tutorial_fast).
```

Comment:
I wonder if this has any effect on the performance of the model: since short or long text may contain different topics, this may make gradients more homogenous compared to regular (fully random) batching. Is there any reference that shows this does not happen? Could you add that here, if one exists? Reading this I get the impression that this might be a trade-off between cost of training and accuracy.

Reply:
Internal tests show that it converged about 2.5x faster with a difference of like 0.05 ppl worse.

It's actually a little more complex than presented here. Namely, for fp16 reasons, we treat all examples the same length up to a divisor of 8. So we actually get a fair amount of stochasticity.